ocaml/idl/schematest.ml

```diff
@@ -0,0 +1,58 @@
+let hash x = Digest.string x |> Digest.to_hex
+
+(* (md5 hash, (major, minor)) *)
+let hash_to_version = [("f3b7df44ff977a0243abc6742b64c418", (5, 700))]
+
+let no_bump hash =
+  Printf.sprintf {|("%s", (%i, %i))|} hash Datamodel_common.schema_major_vsn
+    Datamodel_common.schema_minor_vsn
+
+let minor_bump hash =
+  Printf.sprintf {|("%s", (%i, %i))|} hash Datamodel_common.schema_major_vsn
+    (Datamodel_common.schema_minor_vsn + 1)
+
+let major_bump hash =
+  Printf.sprintf {|("%s", (%i, %i))|} hash
+    (Datamodel_common.schema_major_vsn + 1)
+    0
+
+let schema_hash : string =
+  let hash_of_obj x =
+    Datamodel_types.rpc_of_obj x |> Jsonrpc.to_string |> hash
+  in
+  Datamodel.all_system |> List.map hash_of_obj |> String.concat ":" |> hash
+
+(* Here we test to see if the datamodel has changed. If it has, you
+ * should add the new hash to the [hash_to_version] table *)
+let () =
+  match List.assoc_opt schema_hash hash_to_version with
+  | None ->
+      Printf.eprintf
+        {|
+
+New schema hash='%s'! Add one of the following to the 'hash_to_version' map:
+
+ - '%s' (you didn't modify any fields)
+ - '%s' (you modified a field)
+ - '%s' (you want a new major version)
+
+|}
+        schema_hash (no_bump schema_hash) (minor_bump schema_hash)
+        (major_bump schema_hash) ;
+      failwith "New schema hash"
+  | Some (major, minor) ->
+      if
+        major <> Datamodel_common.schema_major_vsn
+        || minor <> Datamodel_common.schema_minor_vsn
+      then (
+        Printf.eprintf
+          {|
+
+Got known schema hash, but this corresponds to an unexpected schema version. did you bump the schema version unnecessarily?
```

Comment:
This is what usually happens when bumping the major version, it's done when there is a new release, not when there are changes in the schema. (and then the minor is also usually bumped along with it)

Reply:
That's a good point. I think this means we'll want hashes to correspond to multiple versions, instead of just one version

(We could alternatively include the version in the hash, but this creates a chicken vs egg problem like @lindig mentioned!)

Reply:
Ah, I think there is a change needed in this case so it makes sense to exit here with an error since the version in this file doesn't match the one in the datamodel.

I suggest to only change the message to say that these two values are out of sync and must match.